src/Contract/Operation/Matchingable.php

```diff
@@ -0,0 +1,28 @@
+<?php
+
+/**
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace loophp\collection\Contract\Operation;
+
+use Doctrine\Common\Collections\Criteria;
+use Doctrine\Common\Collections\Selectable;
+use loophp\collection\Contract\Collection;
+
+/**
+ * @template TKey
+ * @template T
+ *
+ * @extends Selectable<TKey, T>
+ */
+interface Matchingable extends Selectable
```

Comment:
```suggestion
interface Matchingable
```

@drupol hey, I've looked into the static analysis issues, and here is the problem: we cannot make Matchingable to extend Selectable, because there is no relationship between loophp Collection and Doctrine Collection, so this would break [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle). This is why we're seeing these errors related to covariance reported by PHPStan: https://github.com/loophp/collection/pull/115/files#diff-995edee38ad4f8387e58ebd52c31bcc04c56cc2448d331b1cf5e0b35c57b9efaR9-R19

All errors (both Psalm and PHPStan) disappear if we remove this extending, which I think is what we should do

Reply:
Oh right. I'm mitigated about this.

1. I'm for it because [the return type of Selectable](https://github.com/doctrine/collections/blob/1.6.x/lib/Doctrine/Common/Collections/Selectable.php#L28) is `@psalm-return Collection<TKey,T>` and it doesn't correspond to anything.
2. I'm against it because `Selectable` is something that people are used to manipulate and it would ease the adoption in other places. If we don't do that, I don't see the point of implementing it.

Reply:
unless loophp/collection implements the Doctrine Collection interface, I don't think we should extend `Selectable`. I think there's nothing wrong with having `Matchingable` by itself.

Is your concern related to this use case?

```PHP
function getResults(Selectable $selectable, Criteria $criteria): Doctrine\Collection
{
    // some other code
    return $selectable->matching($criteria);
}
```

For users that are using `matching` like the above, regardless of whether we extend `Selectable` the issue will still be regarding the return type of `getResults`. In their current usage the selectable will return **Doctrine Collection**, however if they try to just pass in **loophp/Collection** into this function, the return type changes to loophp Collection. This might be unexpected and will actually lead to them having to change the return type to `loophp\Collection | Doctrine\Collection` anyway.

If we _don't_ extend Selectable, then this fact is just made a bit more obvious, and they will have to update the function accordingly:

```PHP
/**
* @param Selectable | Matchingable $selectable
*
* @return Doctrine\Collection | loophp\Collection
*/
function getResults(Selectable $selectable, Criteria $criteria)
{
    // some other code
    return $selectable->matching($criteria);
}
```

Reply:
Valid point. I will update accordingly in an hour.

Reply:
Fixed. Thanks!